pcdsdevices/device.py

```diff
@@ -336,3 +341,82 @@ def __copy__(self):
     def create_component(self, instance):
         # Use our hostage component from __set_name__
         return self.copy_cpt.create_component(instance)
+
+
+class GroupDevice(Device):
+    """
+    A device that is a group of components that will act independently.
+
+    This has the following implications:
+    - Components will have no references to this parent device. If
+      accessed and used out of context, the components will be as if
+      they were instantited completely separately.
+    - If a component is staged in a bluesky plan, it will not stage
+      the ``GroupDevice``, and therefore will not stage the entire
+      device tree.
+    - ``GroupDevice`` instances by default do nothing when staged.
+      You can add specific subdevices to the stage list by setting the
+      ``stage_group`` class attribute to a list of components.
+    - Following from the previous point, note that ``GroupDevice``
+      instances cannot process top-level ``stage_sigs``. If you need
+      top-level ``stage_sigs``, you should instead contain them in a
+      subdevice that is not a ``GroupDevice``.
+    - ``GroupDevice`` instances that implement ``set`` are required
+      to specify a ``stage_group`` to help remind you that these classes
+      really do need to stage "something" before scanning. If your
+      movable device really does not need this, you can set ``stage_group``
+      to an empty list.
+    - When represented in typhos, we'll see the GroupDevice screen
+      instead of the default device screens.
+      (Note: at time of writing, this hypothetical ``GroupDevice``
+      ui template does not yet exist).
+    - Certain devices will completely break if we remove their subdevice
+      references: for example, consider the PsuedoPositioner class.
+      For classes like these, we'll need to keep the parent references
+      for the PseudoSingle instances. For the full list of classes that
+      need to retain their ``parent`` attribute, see
+      ``GroupDevice.needs_parent``.
+    """
+    stage_group: list[Component] = None
+    needs_parent: list[OphydObject] = [
+        PseudoSingle,
+        AttributeSignal,
+        PVStateSignal,
+        ]
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Remove references to parent (in this case, self)
+        for cpt_name in self.component_names:
+            cpt = getattr(self, cpt_name)
+            # The following types break without parents
+            if not any(isinstance(cpt, cls) for cls in self.needs_parent):
+                cpt._parent = None
+        if self.stage_group is None:
+            self.stage_group = []
```

Comment:
I got a bit confused reading this because of the line this is attached to, which I think is not related to your comment.

But what you're saying is: in most cases, there is a class-defined `stage_group: list[Component]` and the list is the same object referenced from every instance. And while this does not currently cause a problem, it is good practice to replace the class-shared definition with a new list of the same contents to avoid issues with future code modifying a shared resource.

Is that correct?

Reply:
Yes, you have it exactly. Sorry for the confusion there; it may have been better to comment on the class body list instead.